File: vb/VS2013/netframework/Modules/10.API/35.Object Explorer/MainForm.vb
Imports System.Collections
Imports System.ComponentModel
Imports System.IO

Imports FlexCel.Core
Imports FlexCel.XlsAdapter

Namespace ObjectExplorer
	''' <summary>
	''' Object explorer.
	''' </summary>
	Partial Public Class mainForm
		Inherits System.Windows.Forms.Form

		Public Sub New()
			InitializeComponent()
			ResizeToolbar(mainToolbar)
		End Sub

		Private Sub ResizeToolbar(ByVal toolbar As ToolStrip)

			Using gr As Graphics = CreateGraphics()
				Dim xFactor As Double = gr.DpiX / 96.0
				Dim yFactor As Double = gr.DpiY / 96.0
				toolbar.ImageScalingSize = New Size(CInt(Fix(24 * xFactor)), CInt(Fix(24 * yFactor)))
				toolbar.Width = 0 'force a recalc of the buttons.
			End Using
		End Sub


		#Region "Global variables"
		Private Xls As XlsFile
		#End Region

		Private Sub btnExit_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnExit.Click
			Close()
		End Sub

		Private Sub btnOpenFile_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnOpenFile.Click
			If openFileDialog1.ShowDialog() <> System.Windows.Forms.DialogResult.OK Then
				Return
			End If

			Xls = New XlsFile()
			Xls.Open(openFileDialog1.FileName)


			cbSheet.Items.Clear()
			Dim ActSheet As Integer = Xls.ActiveSheet
			For i As Integer = 1 To Xls.SheetCount
				Xls.ActiveSheet = i
				cbSheet.Items.Add(Xls.SheetName)
			Next i
			Xls.ActiveSheet = ActSheet
			cbSheet.SelectedIndex = ActSheet - 1

			FillListBox()
		End Sub

		Private Sub FillListBox()
			lblObjects.Text = openFileDialog1.FileName
			dataGrid.DataSource = Nothing

			ObjTree.BeginUpdate()
			Try
				ObjTree.Nodes.Clear()

				For i As Integer = 1 To Xls.ObjectCount
					Dim ShapeProps As TShapeProperties = Xls.GetObjectProperties(i, True)
					Dim s As String = "Object " & i.ToString()
					If ShapeProps.ShapeName IsNot Nothing Then
						s = ShapeProps.ShapeName
					End If

					Dim RootNode As New TreeNode(s)
					FillNodes(ShapeProps, RootNode)


					ObjTree.Nodes.Add(RootNode)
				Next i
			Finally
				ObjTree.EndUpdate()
			End Try
		End Sub

		Private Sub FillNodes(ByVal ShapeProps As TShapeProperties, ByVal Node As TreeNode)
			Node.Tag = ShapeProps 'In this simple demo we will use the tag property to store the Shape properties. This is not indented for 'real' use.


			For i As Integer = 1 To ShapeProps.ChildrenCount
				Dim ChildProps As TShapeProperties = ShapeProps.Children(i)
				Dim ShapeName As String = ChildProps.ShapeName
				If ShapeName Is Nothing Then
					ShapeName = "Object " & i.ToString()
				End If
				Dim Child As New TreeNode(ShapeName)
				FillNodes(ChildProps, Child)
				Node.Nodes.Add(Child)
			Next i
		End Sub

		Private Sub btnOpen_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnShowInExcel.Click
			If Xls Is Nothing Then
				MessageBox.Show("There is no open file")
				Return
			End If
			System.Diagnostics.Process.Start(Xls.ActiveFileName)
		End Sub


		Private Sub btnSaveImage_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnSaveAsImage.Click
			If PreviewBox.Image Is Nothing Then
				MessageBox.Show("There is no selected image to save", "Error")
				Return
			End If
			If saveImageDialog.ShowDialog() <> System.Windows.Forms.DialogResult.OK Then
				Return
			End If

			PreviewBox.Image.Save(saveImageDialog.FileName)
		End Sub

		Private Sub btnInfo_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnInfo.Click
			MessageBox.Show("Object Explorer allows to explore inside the objects in an Excel file." & vbLf & "Objects in xls files are hierachily distributed, you can have two objects grouped as a third object, " & "and this hierarchy is shown in the 'Objects' pane at the left. The properties for the selected object are displayed at the 'Object properties' pane.")
		End Sub

		Private Sub ObjTree_AfterSelect(ByVal sender As Object, ByVal e As System.Windows.Forms.TreeViewEventArgs) Handles ObjTree.AfterSelect
			RenderNode(e.Node)
			FillProperties(e.Node)
		End Sub


		Private Sub RenderNode(ByVal Node As TreeNode)
			If Xls Is Nothing Then
				PreviewBox.Image = Nothing
				Return
			End If

			Dim t As TreeNode = Node
			If t Is Nothing Then
				PreviewBox.Image = Nothing
				Return
			End If

			Do While t.Parent IsNot Nothing 'Only root level objects will be rendered.
				t = t.Parent
			Loop

			If t.Index + 1 > Xls.ObjectCount Then
				PreviewBox.Image = Nothing
				Return
			End If

			PreviewBox.Image = Xls.RenderObject(t.Index + 1)


		End Sub

		Private Sub FillProperties(ByVal Node As TreeNode)
			lblObjName.Text = "Name:"
			lblObjText.Text = "Text:"
			Dim Props As TShapeProperties = CType(Node.Tag, TShapeProperties)
			If Props Is Nothing Then
				dataGrid.DataSource = Nothing
				Return
			End If

			Dim ShapeOptions As TShapeOptionList = (TryCast(Node.Tag, TShapeProperties)).ShapeOptions
			If ShapeOptions Is Nothing Then
				dataGrid.DataSource = Nothing
				Return
			End If

			lblObjName.Text = "Name: " & Props.ShapeName
			lblObjText.Text = "Text: " & Convert.ToString(Props.Text)

			Dim ShapeOpts As New ArrayList()
			For Each opt As KeyValuePair(Of TShapeOption, Object) In ShapeOptions
				ShapeOpts.Add(New KeyValue(opt.Key, ShapeOptions))
			Next opt
			dataGrid.DataSource = ShapeOpts
		End Sub

		Private Sub cbSheet_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles cbSheet.SelectedIndexChanged
			Xls.ActiveSheet = cbSheet.SelectedIndex + 1
			FillListBox()
		End Sub

		Private Sub btnStretchPreview_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnStretchPreview.Click
			If btnStretchPreview.Checked Then
				PreviewBox.SizeMode = PictureBoxSizeMode.StretchImage
			Else
				PreviewBox.SizeMode = PictureBoxSizeMode.Normal
			End If

		End Sub

	End Class

	Friend Class KeyValue
		Private FKey As String
		Private FAs1616 As String
		Private FAsLong As String
		Private FAsString As String

		Public Sub New(ByVal aKey As TShapeOption, ByVal List As TShapeOptionList)
			FKey = Convert.ToString(aKey)
			FAs1616 = Convert.ToString(List.As1616(aKey, 0))
			FAsLong = Convert.ToString(List.AsLong(aKey, 0))
			FAsString = List.AsUnicodeString(aKey, "")
		End Sub

		Public Property Key() As String
			Get
				Return FKey
			End Get
			Set(ByVal value As String)
				FKey = value
			End Set
		End Property
		Public Property As1616() As String
			Get
				Return FAs1616
			End Get
			Set(ByVal value As String)
				FAs1616 = value
			End Set
		End Property
		Public Property AsLong() As String
			Get
				Return FAsLong
			End Get
			Set(ByVal value As String)
				FAsLong = value
			End Set
		End Property
		Public Property AsString() As String
			Get
				Return FAsString
			End Get
			Set(ByVal value As String)
				FAsString = value
			End Set
		End Property

		Public Overrides Function ToString() As String
			Return Key
		End Function

	End Class
End Namespace

File: csharp/VS2008/netframework/Modules/10.API/35.Object Explorer/MainForm.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using System.IO;

using FlexCel.Core;
using FlexCel.XlsAdapter;
using System.Collections.Generic;
using System.Diagnostics;

namespace ObjectExplorer
{
    /// <summary>
    /// Object explorer.
    /// </summary>
    public partial class mainForm: System.Windows.Forms.Form
    {

        public mainForm()
        {
            InitializeComponent();
            ResizeToolbar(mainToolbar);
        }

        private void ResizeToolbar(ToolStrip toolbar)
        {

            using (Graphics gr = CreateGraphics())
            {
                double xFactor = gr.DpiX / 96.0;
                double yFactor = gr.DpiY / 96.0;
                toolbar.ImageScalingSize = new Size((int)(24 * xFactor), (int)(24 * yFactor));
                toolbar.Width = 0; //force a recalc of the buttons.
            }
        }


        #region Global variables
        XlsFile Xls;
        #endregion

        private void btnExit_Click(object sender, System.EventArgs e)
        {
            Close();
        }

        private void btnOpenFile_Click(object sender, System.EventArgs e)
        {
            if (openFileDialog1.ShowDialog() != DialogResult.OK) return;

            Xls = new XlsFile();
            Xls.Open(openFileDialog1.FileName);


            cbSheet.Items.Clear();
            int ActSheet = Xls.ActiveSheet;
            for (int i = 1; i <= Xls.SheetCount; i++)
            {
                Xls.ActiveSheet = i;
                cbSheet.Items.Add(Xls.SheetName);
            }
            Xls.ActiveSheet = ActSheet;
            cbSheet.SelectedIndex = ActSheet - 1;

            FillListBox();
        }

        private void FillListBox()
        {
            lblObjects.Text = openFileDialog1.FileName;
            dataGrid.DataSource = null;

            ObjTree.BeginUpdate();
            try
            {
                ObjTree.Nodes.Clear();

                for (int i = 1; i <= Xls.ObjectCount; i++)
                {
                    TShapeProperties ShapeProps = Xls.GetObjectProperties(i, true);
                    string s = "Object " + i.ToString();
                    if (ShapeProps.ShapeName != null) s = ShapeProps.ShapeName;

                    TreeNode RootNode = new TreeNode(s);
                    FillNodes(ShapeProps, RootNode);


                    ObjTree.Nodes.Add(RootNode);
                }
            }
            finally
            {
                ObjTree.EndUpdate();
            }
        }

        private void FillNodes(TShapeProperties ShapeProps, TreeNode Node)
        {
            Node.Tag = ShapeProps; //In this simple demo we will use the tag property to store the Shape properties. This is not indented for 'real' use.


            for (int i = 1; i <= ShapeProps.ChildrenCount; i++)
            {
                TShapeProperties ChildProps = ShapeProps.Children(i);
                string ShapeName = ChildProps.ShapeName;
                if (ShapeName == null) ShapeName = "Object " + i.ToString();
                TreeNode Child = new TreeNode(ShapeName);
                FillNodes(ChildProps, Child);
                Node.Nodes.Add(Child);
            }
        }

        private void btnOpen_Click(object sender, System.EventArgs e)
        {
            if (Xls == null)
            {
                MessageBox.Show("There is no open file");
                return;
            }
            using (Process p = new Process())
            {               
                p.StartInfo.FileName = Xls.ActiveFileName;
                p.StartInfo.UseShellExecute = true;
                p.Start();
            }              
        }


        private void btnSaveImage_Click(object sender, System.EventArgs e)
        {
            if (PreviewBox.Image == null)
            {
                MessageBox.Show("There is no selected image to save", "Error");
                return;
            }
            if (saveImageDialog.ShowDialog() != DialogResult.OK) return;

            PreviewBox.Image.Save(saveImageDialog.FileName);
        }

        private void btnInfo_Click(object sender, System.EventArgs e)
        {
            MessageBox.Show("Object Explorer allows to explore inside the objects in an Excel file.\n" +
                "Objects in xls files are hierachily distributed, you can have two objects grouped as a third object, " +
                "and this hierarchy is shown in the 'Objects' pane at the left. The properties for the selected object are displayed at the 'Object properties' pane.");
        }

        private void ObjTree_AfterSelect(object sender, System.Windows.Forms.TreeViewEventArgs e)
        {
            RenderNode(e.Node);
            FillProperties(e.Node);
        }


        private void RenderNode(TreeNode Node)
        {
            if (Xls == null)
            {
                PreviewBox.Image = null;
                return;
            }

            TreeNode t = Node;
            if (t == null)
            {
                PreviewBox.Image = null;
                return;
            }

            while (t.Parent != null) t = t.Parent;  //Only root level objects will be rendered.

            if (t.Index + 1 > Xls.ObjectCount)
            {
                PreviewBox.Image = null;
                return;
            }

            PreviewBox.Image = Xls.RenderObject(t.Index + 1);


        }

        private void FillProperties(TreeNode Node)
        {
            lblObjName.Text = "Name:";
            lblObjText.Text = "Text:";
            TShapeProperties Props = (TShapeProperties)Node.Tag;
            if (Props == null)
            {
                dataGrid.DataSource = null;
                return;
            }

            TShapeOptionList ShapeOptions = (Node.Tag as TShapeProperties).ShapeOptions;
            if (ShapeOptions == null)
            {
                dataGrid.DataSource = null;
                return;
            }

            lblObjName.Text = "Name: " + Props.ShapeName;
            lblObjText.Text = "Text: " + Props.Text;

            ArrayList ShapeOpts = new ArrayList();
            foreach (KeyValuePair<TShapeOption, object> opt in ShapeOptions)
            {
                ShapeOpts.Add(new KeyValue(opt.Key, ShapeOptions));
            }
            dataGrid.DataSource = ShapeOpts;
        }

        private void cbSheet_SelectedIndexChanged(object sender, System.EventArgs e)
        {
            Xls.ActiveSheet = cbSheet.SelectedIndex + 1;
            FillListBox();
        }

        private void btnStretchPreview_Click(object sender, EventArgs e)
        {
            if (btnStretchPreview.Checked)
                PreviewBox.SizeMode = PictureBoxSizeMode.StretchImage;
            else
                PreviewBox.SizeMode = PictureBoxSizeMode.Normal;

        }

    }

    class KeyValue
    {
        private string FKey;
        private string FAs1616;
        private string FAsLong;
        private string FAsString;

        public KeyValue(TShapeOption aKey, TShapeOptionList List)
        {
            FKey = Convert.ToString(aKey);
            FAs1616 = Convert.ToString(List.As1616(aKey, 0));
            FAsLong = Convert.ToString(List.AsLong(aKey, 0));
            FAsString = List.AsUnicodeString(aKey, "");
        }

        public string Key { get { return FKey; } set { FKey = value; } }
        public string As1616 { get { return FAs1616; } set { FAs1616 = value; } }
        public string AsLong { get { return FAsLong; } set { FAsLong = value; } }
        public string AsString { get { return FAsString; } set { FAsString = value; } }

        public override string ToString()
        {
            return Key;
        }

    }
}
